File: src/main/java/com/bdev/hogwarts_api/rest_service/student_actions/StudentAttendanceRestServiceImpl.java
package com.bdev.hogwarts_api.rest_service.student_actions;

import acropollis.municipali.security.common.dto.MunicipaliUserInfo;
import com.bdev.hogwarts_api.data.dto.student.StudentAttendance;
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException;
import com.bdev.hogwarts_api.service.student.StudentService;
import com.bdev.hogwarts_api.service.student_attendance.StudentAttendanceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class StudentAttendanceRestServiceImpl implements StudentAttendanceRestService {
    @Autowired
    private StudentService studentService;

    @Autowired
    private StudentAttendanceService studentAttendanceService;

    @Transactional(readOnly = true)
    @Override
    public List<StudentAttendance> getAllAttendances(MunicipaliUserInfo userInfo) {
        return studentAttendanceService.getAllAttendances();
    }

    @Transactional(readOnly = true)
    @Override
    public List<StudentAttendance> getAttendances(MunicipaliUserInfo userInfo, long studentId) {
        if (!studentService.getStudentById(studentId).isPresent()) {
            throw new HttpEntityNotFoundException("Student with id '%d' does not exist", studentId);
        }

        return studentAttendanceService.getAttendances(studentId);
    }

    @Transactional
    @Override
    public long addAttendance(MunicipaliUserInfo userInfo, StudentAttendance attendance) {
        if (!studentService.getStudentById(attendance.getStudentId()).isPresent()) {
            throw new HttpEntityNotFoundException("Student with id '%d' does not exist", attendance.getStudentId());
        }

        return studentAttendanceService.addAttendance(attendance);
    }

    @Transactional
    @Override
    public void deleteAttendance(MunicipaliUserInfo userInfo, long attendanceId) {
        if (!studentAttendanceService.exists(attendanceId)) {
            throw new HttpEntityNotFoundException("Student attendance with id '%d' does not exist", attendanceId);
        }

        studentAttendanceService.deleteAttendance(attendanceId);
    }
}

File: src/main/java/com/bdev/hogwarts_api/rest_service/student_actions/StudentAttendanceRestServiceImpl.kt
package com.bdev.hogwarts_api.rest_service.student_actions

import acropollis.municipali.security.common.dto.MunicipaliUserInfo
import com.bdev.hogwarts_api.data.dto.student.StudentAttendance
import com.bdev.hogwarts_api.exceptions.http.HttpEntityNotFoundException
import com.bdev.hogwarts_api.service.student_legacy.StudentService
import com.bdev.hogwarts_api.service.student_attendance.StudentAttendanceService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class StudentAttendanceRestServiceImpl : StudentAttendanceRestService {
    @Autowired
    private lateinit var studentService: StudentService

    @Autowired
    private lateinit var studentAttendanceService: StudentAttendanceService

    @Transactional(readOnly = true)
    override fun getAllAttendances(userInfo: MunicipaliUserInfo): List<StudentAttendance> {
        return studentAttendanceService.getAllAttendances()
    }

    @Transactional(readOnly = true)
    override fun getAttendances(userInfo: MunicipaliUserInfo, studentId: Long): List<StudentAttendance> {
        if (studentService.getStudentById(studentId) == null) {
            throw HttpEntityNotFoundException("Student with id '%d' does not exist", studentId)
        }

        return studentAttendanceService.getAttendances(studentId)
    }

    @Transactional
    override fun addAttendance(userInfo: MunicipaliUserInfo, attendance: StudentAttendance) {
        if (studentService.getStudentById(attendance.studentId) == null) {
            throw HttpEntityNotFoundException("Student with id '%d' does not exist", attendance.studentId)
        }

        studentAttendanceService.addAttendance(attendance)
    }

    @Transactional
    override fun deleteAttendance(userInfo: MunicipaliUserInfo, studentId: Long, startTime: Long) {
        studentAttendanceService.deleteAttendance(
                studentId = studentId,
                startTime = startTime
        )
    }
}
